pkg/proto/proto.go

```diff
@@ -76,7 +233,48 @@ type Header struct {
 	Magic           uint32
 	ContentID       PeerMessageID
 	payloadLength   uint32
-	PayloadChecksum [headerChecksumLen]byte
+	PayloadChecksum [payloadChecksumSize]byte
+}
+
+func NewHeader(contentID PeerMessageID, body []byte) (Header, error) {
+	bl, err := safecast.ToUint32(len(body))
+	if err != nil {
+		return Header{}, fmt.Errorf("failed to create header: %w", err)
+	}
+	msgLen := msgMagicSize + msgTypeSize + payloadLenSize // For empty Header.
+	cs := [payloadChecksumSize]byte{}
+	if bl > 0 {
+		msgLen = msgMagicSize + msgTypeSize + payloadLenSize + payloadChecksumSize + bl
+		dig, fhErr := crypto.FastHash(body)
+		if fhErr != nil {
+			return Header{}, fmt.Errorf("failed to create header: %w", fhErr)
+		}
+		copy(cs[:], dig[:payloadChecksumSize])
+	}
+	return Header{
+		Length:          msgLen,
+		Magic:           headerMagic,
+		ContentID:       contentID,
+		payloadLength:   bl,
+		PayloadChecksum: cs,
+	}, nil
+}
+
+// Validate checks the header for correctness. It checks the magic number, ContentID, and lengths.
+// Returns an error with a description of the problem.
+func (h *Header) Validate(contentID PeerMessageID) error {
+	if h.Magic != headerMagic {
+		return fmt.Errorf("invalid header: wrong magic: want %x, have %x", headerMagic, h.Magic)
+	}
+	if h.ContentID != contentID {
+		return fmt.Errorf("invalid header: wrong ContentID: want %x, have %x", contentID, h.ContentID)
+	}
+	// h.Length is the length of the message after the MSG_LEN field itself. So, we need to add 4 bytes to check
+	// the total length of the message
+	if act, exp := h.Length+uint32Size, h.HeaderLength()+h.payloadLength; act != exp {
```

Comment:
I think `h.Length+uint32Size` value should be `exp`. ([previous issue](https://github.com/wavesplatform/gowaves/pull/1579#discussion_r1922815661))

Reply:
Nope, we are checking the `h.Length` but I moved the 4 bytes on the Length to the expected part. Actually, there is 2 variables `h.Length` and `h.payloadLength`, so we can check the first one or the second. It depends on point of view.

Reply:
Agree